### ISSUE TITLE
Add support for profiles

### ISSUE THREAD

**Issue by hssm** [ankidroid/Anki-Android#2545]
> [!NOTE]
> * Moved to https://github.com/ankidroid/Anki-Android/issues/18117

----

Parallel builds simulate profiles. They don't auto-update, but you can find them:

* Latest Stable https://github.com/ankidroid/anki-android/releases/latest
* Latest Alpha https://github.com/ankidroid/anki-android/releases/

_edited by @david-allison_

----

Originally reported on Google Code with ID 1643

```
Anki 2.0 supports having different profiles for different users.
This also allows a user to split their decks between different collections.

I built a prototype of this and it seems quite simple.
The prototype is based on using different preference files.
```

Reported by `flerda` on 2013-01-26 01:18:26

**Comment by MatthiasKauer** [comment]
> We're currently recommending users who want this feature to do this:
https://code.google.com/p/ankidroid/wiki/FAQ#How_to_use_different_Anki_profiles

That page no longer exists. Where should we look now?

**Comment by flerda** [comment]
https://github.com/ankidroid/Anki-Android/wiki/FAQ#how-to-use-different-anki-profiles

**Comment by timrae** [comment]
I've made a one-time release of 3 parallel builds for v2.5beta16:
https://github.com/ankidroid/Anki-Android/releases/tag/v2.5beta16

These will probably not be updated in the future, but I've created a bash script that you can use to [easily make your own parallel builds](https://github.com/ankidroid/Anki-Android/wiki/Development-Guide#making-parallel-builds) in the future.

**Comment by langug** [comment]
This is really an important feature. I hope it will soon be integrated in the normal version of AnkiDroid.

**Comment by michalpravda** [comment]
For me it is by far the most important missing feature, don't know how to "vote it up"

**Comment by timrae** [comment]
There isn't a way to vote issues up and open source developers don't tend to work on features based on their popularity anyway (rather on issues that are fun or personally useful for them). If the feature is really important to you, I'd suggest offering a bounty to give developers a financial incentive.

**Comment by michalpravda** [comment]
Thanks for the reply, I have in the meantime experimented with your parallel builds and after some struggling made that work. The most (and the only) non-obvious thing being the necessity to change the collection location in each (well, in "n-1") parallel build.
I'm willing to pay about 5$ for the parallel builds and another 10 as a thank you for the application, but:
1. don't know if it would please you or you would feel offended since it is such a small amount
2. don't know how to easily do it (have Paypal, card - have been searching a way for ten minutes now)

**Comment by timrae** [comment]
@michalpravda 

It's not possible to donate directly to AnkiDroid, but if there is a specific feature that you would like to incentivize developers to work on, there are sites where you can set a "bounty". Even if you can only pledge a small amount, others may also pledge something which together could form enough to motivate a developer.

I've never used it before, but https://www.bountysource.com/ appears to be one popular website for doing this.

**Comment by mikehardy** [comment]
@timrae are you still interested in merging changes to support multiple profiles in AnkiDroid? I'm going to do a parallel build first (I could do 2 or 3 and publish those as well if there is interest...) but since it would affect UI (for switching) plus lots of storage stuff I don't even want to think much about it if there's no interest...

**Comment by timrae** [comment]
@Mike Hardy

Yes, definitely interested in merging profile support if done right!

**Comment by michalpravda** [comment]
@Mike Hardy

I'm interested as well. I have three parallel 2.5beta16 on my phone. Used
weekly by me and my kids. Would be interested in more up to date parallel
builds if the profile feature didn't work out.

2017-12-12 15:07 GMT+01:00 Tim Rae <notifications@github.com>:

> @Mike Hardy
>
> Yes, definitely interested in merging profile support if done right!
>

**Comment by mikehardy** [comment]
@michalpravda As mentioned I'm first trying to do fresh parallel builds - it's not the most useful work but I'm becoming familiar with mobile development in general and the AnkiDroid codebase specifically at the same time so it's useful for me. Unfortunately I ran into some problems immediately - the parallel builds script had an interaction with current the AndroidStudio build system, it didn't change the storage directory (leading to confusion others noted), and while those were easy then when I installed it it didn't work out of the box - as if some part of the system were still referring to "AnkiDroid" instead of "AnkiDroid.A" for instance. My progress is available on a topic branch on my local fork, if any one else has interest as well - my work on it will be pretty intermittent - https://github.com/mikehardy/Anki-Android/tree/parallel1

**Comment by michalpravda** [comment]
@mikehardy - I have no experience with mobile development, but I do
remember that it took me some time to configure the builds I had received
from @TimRae. The problem was that all instances pointed to the same path
for storing the collection's data. Once I manually set it to different
paths (ankidroid - settings - ?advanced? - ?path to collection? (I
use localized version and now just guessing the menu entries)) it began to
work flawlessly. Maybe that's your missing piece?!

2017-12-13 17:54 GMT+01:00 Mike Hardy <notifications@github.com>:

**Comment by mikehardy** [comment]
Yeah I saw that comment earlier about the confusion there and updated the parallel build script to change that storage directory in each build as well, I'm having some other thing, I'm sure it is basic and Ill learn something...

**Comment by timrae** [comment]
I don't think that's something we'd want to change upstream in the script,
but if it helps you to figure out the code then by all means.

**Comment by mikehardy** [comment]
Hmm - Looking at the effects of different storage locations will definitely help me figure out persistence as I learn where everything is stored in order to implement separate sync accounts / decks, but I’m curious on the stance of not merging a storage directory change in the parallel builds script upstream - only because I can only think of one use case one would want parallel builds, and that’s to have separate decks etc so that they can simulate different profiles with different users. If (big if) that is the only use case, than it makes sense for the parallel builds to also have parallel storage. But, I could  be wrong for a number of reasons - I could be missing a use case, violating principle of least surprise, or simply proposing something that smells bad for reasons I’m unaware

**Comment by timrae** [comment]
@mikehardy 

> I can only think of one use case one would want parallel builds, and that’s to have separate decks etc so that they can simulate different profiles with different users

Yes that's a fair assumption, I guess was I just imagining some horrible hack, but it doesn't actually look that difficult to do it properly.

**Comment by mikehardy** [comment]
I have created a [pull request](https://github.com/ankidroid/Anki-Android/pull/4751) with the changes from [my fork's parallel1 branch](https://github.com/mikehardy/Anki-Android/tree/parallel1) that seem to correctly generate backwards-compatible parallel builds with separate storage directories for fresh installs now. I have generated up to date packages on my local fork from this work and made a release of them.

If others are interested in testing up-to-date parallel builds and are willing to accept the consequences of testing untrusted forked alpha code ;-), please check them out:
https://github.com/mikehardy/Anki-Android/releases/tag/v2.9alpha16

They aren't perfect but I noted the issues on the release page, and they do work for me.

However, please please do not log any issues or otherwise create any work for @timrae  or other real developers as a result of them. They are convenience only for others on this issue, and I'll pull them if they cause any problems whatsoever, or if the main project here doesn't like them

That said, if they work or don't and you have useful feedback I'd be interested to know.

Next up for me is to think about actually implementing multiple profiles.

**Comment by mikehardy** [comment]
Future implementation of this should take into account [grouping notification channels by profile](https://developer.android.com/training/notify-user/channels#CreateChannelGroup) now that we support channels: #4922

**Comment by madmanpierre** [comment]
> @timrae are you still interested in merging changes to support multiple profiles in AnkiDroid? I'm going to do a parallel build first (I could do 2 or 3 and publish those as well if there is interest...) but since it would affect UI (for switching) plus lots of storage stuff I don't even want to think much about it if there's no interest...

I currently don't use the Android, I just use my laptop. I have 4 kids in the learning to read level, and I would love to be able to use my Android phone reasonably. I tried the parallel build but really didn't try hard due to not quite understanding and decided to wait until it was an official feature, and just use my laptop for now.

**Comment by mikehardy** [comment]
@madmanpierre I can understand the fear of a rug being pulled from you, but I can say (as the person currently publishing modern parallel releases) that the "hack" really can't matter much. Either it syncs successfully to a separate AnkiWeb account or it doesn't, right? So in the future if it's implemented any other way and sync (one of the key features in this app) works, you'll just re-sync.

The only problem I could see for you is that I only do 3 extra install builds when I generate parallel releases, and you have 4 kids! Maybe I could add one... :-)

Literally all the parallel build is doing is making a separate Android app package name, so you can maintain a separate preference file, then in that preferences you define a separate AnkiDroid directory, thereby keeping your collections separate.

I only have one kid, but I use it for exactly this case.

**Comment by madmanpierre** [comment]
Thanks Mike. If you could add one more that would be nice ! Lots of useless but MAYBE interesting info:

I've been using Anki personally for over 10 years (and did give my cash donation, it was well worth it). I've been using it for learning French, Spanish, computer commands and every other misc 'minor' thing I want to memorize. I have 10's of thousands of cards in my active decks. (and I have them all intermixed by grouping the decks to review, I find it helps get my learning solid when it jumps between French, Spanish, Linux, Windows, Misc Facts etc.., during my, normally 30 minute a day period, adding about 5 'facts to memorize' a day.) 

My wife and I are foster parents. We have had 4 kids in our care for several years (all since birth), the oldest is grade 1, the youngest a couple years behind, the youngest ones wanted to use the 'same thing' that my grade one daughter is using to learn how to read. 

In my kid's case, the front and back of cards are identical because I am the arbitrator of whether they have read and understood (there/their, by/bye etc..,) the word correctly. But it is SO nice to let Anki worry about the timing of the cards - so much better than flash cards.

As a specific for how good it was. My 6 year old was working with Mom and adult big sister for almost a month and, using flash cards, had learned to read only two words by sight. I offered to take over, put all the sight words into Anki, and my daughter learned 40 words in less than a week, and has been learning now at 4x's the rate the school is asking for - and she uses (with me) Anki only about 5 minutes a day, been doing it for a month now, and she has never made a mistake on a word after it gets left for 7 days, so I KNOW she is learning them solid. 

So … if that motivates you to add a 4th extra build … I assure you the facts are all true! In any event, over the holidays I'll set up at least the 3 extras then and feel comfortable sitting on the rug. Thanks.

**Comment by mikehardy** [comment]
@madmanpierre I'm a sucker for any parent putting in the work :-). I'm having roughly the same experience with my 7 year old - regular flash cards? Oh the pain. AnkiDroid? Just a few minutes a day while waiting in line or something. Huge progress.

I increased the parallel build count to 5, just in case. Happy holidays

https://github.com/mikehardy/Anki-Android/releases

Eventually I'm going to implement the feature for real but it's a functional band-aid at least, though you will have to update them manually. Also, yes they are alphas but I use them daily - we are very close to a 2.9 beta then release cycle.

**Comment by michalpravda** [comment]
@madmanpierre - Impressive. .... how do you get/choose those 'facts to
memorize'?

I use Anki solely for teaching my kids English as a second language. My
kids mostly arbitrate themselves (the 11y daughter cheats sometimes, the 7y
son never). I've got anki[droid] installed on 7 devices (2pc,1 tablet, 4
phones) and everything works almost perfectly. The only thing not working
as I would imagine is autosyncing after the lesson. Kids usually finish
their batches then press the home button and run a game. Anki gets
"backgrounded" and the sync doesn't happen. Sometimes they even "kill all
running apps". It's no dealbreaker, though.

M

pá 21. 12. 2018 v 6:08 odesílatel peter horwood aka Madman Pierre <
notifications@github.com> napsal:

**Comment by madmanpierre** [comment]
>how do you get/choose those 'facts to memorize'?

Well, first of all, I've used Anki for enough years to have a 'gut feel' for what types of things Anki can help me learn and what it can't realistically help with. I also have a 'gut feel' for what types of things I will need 'help' memorizing - for example, words/phrases/acronyms I will need occasionally but not enough to 'naturally' memorize them.

From there I have 2 main strategies:

  *
proactive (when I'm getting into a new area in computers - I run a multinational computer company, I'm the tech lead - I'll go and do some study and put in a few dozen or more words etc.., that it is obvious I'll need to easily work with.)
  *
reactive (when I come across a word etc.., the first time that I think fits my profile, and when I come across it the 3rd time even if it didn't seem to fit the proile.)

________________________________

**Comment by sudomain** [comment]
I'm working on a Anki related profile issue and have a question. I know I can change the AnkiDroid directory via a setting in the "Advanced" tab in preferences. Is the AnkiDroid directory the same as the per profile directories that Anki desktop creates (e.g. ~/.local/share/Anki2/User 1 )? At the very least it seems similar in structure (collection.anki2, collection.media, etc.) 

I'm only asking because I have been able to install Anki 2.1 on Android via a Proot environment in Termux. If I can tell AnkiDroid to use my Anki profile as the AnkiDroid directory then I'll save a fair amount of space on my device since I wont need two copies of my collection. I also don't want to corrupt my collection though.

**Comment by mikehardy** [comment]
I wouldn't trust it - the databases themselves are similar enough they should be identical, but each system expects to own the files, and at least for AnkiDroid there are other files in the directory besides just the collection. That said, if you wanted to make a test profile to see how it works, that could be interesting

**Comment by sudomain** [comment]
After a few tests (creating decks, adding notes, etc.) it seems ok to have them access the same database / profile directory, though there could be hidden nasty things going on with my collection now. The main concern is when the collection is open in the desktop version, then AnkiDroid is opened, AnkiDroid will report that the database is locked and prompt the user that the collection may be corrupt. Closing either Anki or AnkiDroid then running a database check in the other program will result in no errors. Any decks or notes that were created since the last database update (updates happen at app close?) will still be in the collection. Kudos to Anki and AnkiDroid for making such a resilient system.

Here is a screenshot of a deck and notes that I created in Anki desktop before purposely launching AnkiDroid and causing an apparent collection corruption. Note that both Anki and AnkiDroid report a corruption, but neither report errors when checking the database.


![Screenshot_2019-11-19-17-23-03](https://user-images.githubusercontent.com/30786326/69193384-910bf780-0af4-11ea-8de1-8bca14db3db1.png)

**Comment by mikehardy** [comment]
Fascinating! I think parallel releases and the native interface are probably more pleasant but simply knowing proot exists is interesting. And if a database check and media check pass, I'd be reasonably confident things are fine. At the database level it is not *that* complicated of a system.

**Comment by sudomain** [comment]
That's great to hear about the database check. I would definitely prefer to have the same note creation workflow on AnkiDroid as I have on Anki. Unfortunately even if all of the Anki addons I use were ported to AnkiDroid's addon system, I would still have to contend with Android's weird use of the right click button on hardware mice (in most cases, it interprets it as the back button). Would AnkiDroid welcome an issue/PR that would change how the app interprets right clicks on hardware mice? For now, I luckily have a VNC viewer app that sends the appropriate right clicks and keyboard shortcuts to my Anki proot.

**Comment by p3l4h0** [comment]
I mainly use ios and it works wonderfully. Also the parallel releases of @mikehardy work great for android. But here as an incentive if someone wants to participate (maybe@michalpravda):
https://www.bountysource.com/issues/25121039-add-support-for-profiles

**Comment by sudomain** [comment]
I found #3114 while looking for something else. The two features seem very related. `/sdcard/AnkiDroid` is basically a profile without the surrounding Anki root directory a.k.a. `$ANKI_BASE` on desktop. Neither are features I personally have the time to work on, but whoever does will be able to close 2 issues imo.

**Comment by Akshay0701** [comment]
## Problem 

- If the user wants to maintain many profiles in one device, then there is no option for that in AnkiDroid in one app
- AnkiDesktop have switch profile feature for adding, removing, opening user.

## Data 
  We will store a list of profiles in shared preferences 

## Design 
- new user opens app then user will login through normal login page , and that user details will store in default profile option.
- if user want to create another profile ,then he have to navigate Settings->AnkiDroid ->SwtichUser ,then user can perform operation like add user ,select user ,delete user.
- after adding new profile in list and selecting profile then login page will open for user to login and that info will save to selected profile option (so that in  future if user again select this profile option then it will directly login without asking for all process again).
- If user select already login profile then It will directly take user to HomePage
- After successfully switch, Deckpicker activity will automatically start sync method to resolve conflict between two different account data
<img src="https://user-images.githubusercontent.com/52353967/115337800-48f87500-a1bf-11eb-9bc2-bf170541a5ed.jpg" width="500" height="600" />

**Comment by Akshay0701** [comment]
Let me know, should I implement this feature using this approach?

**Comment by david-allison-1** [comment]
> Let me know, should I implement this feature using this approach?

We should hold off on this feature until #5304 has been implemented. 

Both are going to cause large changes to the codebase regarding storage, and #5304 is more important due to a Google-imposed deadline

**Comment by Akshay0701** [comment]
Yes david i understood , i am just implementing for myself and learning , I am not going to raise pr of this till this open

**Comment by minustimesplus** [comment]
> I agree real profile support is important - and you are right David it is literally the reason I started hacking on this app at all hahaha. One of these days it'll get there in the meantime I'm glad you enjoy the hacky but usable parallel install workaround I made. I do use it myself (for my child's collection) so it's guaranteed to be available until real profile support comes in. Cheers

Hi! Any updates on this?

**Comment by david-allison** [comment]
I'd be willing to take this on, I was planning for it some time in the new year anyway

**Comment by david-allison** [comment]
Draft of my thoughts:

<details>

<summary>Anki's screen</summary>

![Screenshot 2024-01-02 at 13 15 03](https://github.com/ankidroid/Anki-Android/assets/62114487/3b4f2566-1c40-44db-8d8f-6d96a008c485)

</details>

**Functionality**
* A profile does not require an AnkiWeb account
* A profile can be opened
  * Each profile contains a separate `Collection` and set of `SharedPreferences` 
* A user will be able to add a named profile. 
  * Preferences may be copied from another profile 
* A user will be able to rename an existing profile
* A user will be able to delete a profile
* We will not implement 'Downgrade & Quit/Open Backup' here. The Deck Picker contains 'Open Backup' already, and this screen will not be used by the majority of users.

**Defaults**
* We have a 'default' profile, which will launch on open
* The default is the last-opened profile
* The 'open profile' may be shown on launch instead

**Unknown**
* How do we handle widgets?
* How do we handle WebView cookies
   * Specifically: Shared Decks -> AnkiWeb login
* How do we handle app shortcuts
* How do we handle 'Manage Data'
* How do we handle the API
* Profile names may contain special characters. Should the name map to a folder name?
* Our current folder is `AnkiDroid`

**Constraints**
* The standard onboarding process should not be affected, neither will the default user experience. Profiles are an extra
* We should not modify the navigation drawer, as this is soon to be removed

**Comment by BrayanDSO** [comment]
> How do we handle WebView cookies
Specifically: Shared Decks -> AnkiWeb login

Share it, I guess

> How do we handle app shortcuts

Maybe add the profile name to each one? (I have no idea if that is possible) -> Add (David), Study (Arthur)

> Profile names may contain special characters. Should the name map to a folder name?

Mimic Anki desktop behavior or allow any characters in the profile name but strip the unallowed in the folder name

> Our current folder is AnkiDroid

For app private builds, doesn't seem to be a problem

For full builds, ideally the profile should be in `AnkiDroid/profile name` but `AnkiDroid/` already has the collection. The easiest solution I see is keeping the old `AnkiDroid/` folder and if the user tries to create a new profile, default it to a new directory called `AnkiDroid profiles` or something better. For new installs, we can default to this new folder

**Comment by david-allison** [comment]
I specifically chose `AnkiDroid` as a name so we're not writing to the root of `com.ichi2.anki/files` in the private directory (so `AnkiDroid` can be the default profile)

For the public directory, `AnkiDroid Profiles` seems ideal

>> How do we handle app shortcuts
> Maybe add the profile name to each one? (I have no idea if that is possible) -> Add (David), Study (Arthur)

I believe this is workable. Our shortcuts are dynamic.
* + We want a per-profile setting to determine which icons (if any) appear

----

* Widgets: probably 'unsupported' for now and only works on the current profile
* API: probably the same

**Comment by BrayanDSO** [comment]
> Widgets: probably 'unsupported' for now and have them tied to the current profile

For the add note widget, I don't suggest any work because it should deprecated when minSdk reaches Nougat and everyone can use shortcuts.

For the reviews shortcut, it needs a redesign so it can have some configurable aspect to distinct it from other instances of it (a color, icon, emoji, text, whatever)

> I think we want to isolate the WebView cookies if at all possible

Now that I thought a little bit more, I agree

**Comment by david-allison** [comment]
API 28+: [`WebView.setDataDirectorySuffix`](https://developer.android.com/reference/android/webkit/WebView#setDataDirectorySuffix(java.lang.String))

**Comment by david-allison** [comment]
**Pseudocode**

A profile will be of the format:

```kotlin
data class ProfileData(val identifier: String, var name: String) {
    val webViewDataDirectory get() = identifier
    fun preferences(context: Context) = context.getSharedPreferences(
           name = identifier,            
           mode = getDefaultSharedPreferencesMode(),
        )
    fun rename(name: String) = TODO()
    fun select() = TODO()
    fun delete() = TODO("do we also delete data?")
    companion object {
         fun listAll(): List<ProfileData> { }   
    }
}
```

`identifier` must NOT contain path separators (`setDataDirectorySuffix`)
Proposal: current unix timestamp in ms

----

A file containing `List<ProfileData>` should exist, for use on the 

* It would be inefficient to store the `name` inside `ProfileData.preferences()`
  * O(n) 
  * A large amount of XML needs to be read for one field
  * If an issue occurs with a profile, the name is lost
  * But it would be very unusual to have more than 50 profiles, even as a developer
  * architecture is a better argument than speed

Realistic options for storage of the ProfileCollection:

1. in the current `SharedPreferences` using `StringSet` + concatenated string
   * As we've done previously
   * Fast
2. In a new `SharedPreferences`
    * slightly faster than the above 
3. Jetpack Data Store
    * Strongly Typed
    * Async
    * Slower

----

**TODO:** 

* What happens when a profile is deleted?
  * Offer to permanently delete linked collection folder + backups?
    * [Security] if AnkiDroid is acting as a file manager, ensure you can't use this to delete an arbitrary folder
* Can two profiles have the same name?
  * Technically, yes, but block it  
* AnkiDroid needs to know the current profile and load this efficiently (blocker on hot startup path)
* Should the current profile name be visible, are there latency concerns if it's on the hot path?

**Comment by david-allison** [comment]
I'm not going to have capacity to take this through for the 2.19 series

**Comment by Arthur-Milchior** [comment]
https://github.com/user-attachments/assets/4b8bbd4a-7ba7-4fc1-9672-ed8baff3d564
Here is the way the feature work on ankimobile

**Comment by Arthur-Milchior** [comment]
https://github.com/user-attachments/assets/f485cdcb-27c6-4248-9f33-039fca7a3b0f
Here is more usage

**Comment by Arthur-Milchior** [comment]
I believe this issue is far too long. With a lot of discussion that is not relevant anymore, even if it's touching.
I created #18117 where I copy/pasted everything relevant, and suggest we abandon this issue. This way, any GSoC candidate don't have to read 10 years of history and can concentrate on what is actually relevant for their proposal. And we can more easily discuss the details of the implementation
